### PR TITLE
Remove admin::user#destroy endpoint duplication + "admin" specs template proposal

### PR DESCRIPTION
**Pull request by scouillard**
## Description

Remove the duplicated user#destroy endpoint.

**Propose a new template for "admin" specs:**
In specs, it can be difficult to differentiate regular user specs from "admin" users specs.
1. The roles/permissions are set on top of the resource spec files (this can be moved in another module eventually)

```
  let(:<permission>_role) { create(:role) }
  let(:<permission>_permission) { create(:permission, name: '<permission>') }
  let(:<permission>_role_permission) do
    create(:role_permission,
           role: <permission>_role,
           permission: <permission>_permission,
           value: 'true',
           provider: 'greenlight')
  end
```

2. A context is created inside the `describe <endpoint>`

```
describe 'some#endpoint' do
  it 'does something' do 
    test something
  end

  context 'user with <permission> is doing something' do
     before do 
       <permission>_role_permission ## instantiate the role_permission
       user.update!(role: <permission>_role) ## assign current_user the permission
     end

     it 'does something' do 
       test something
     end 
end
```

### CODE REVIEW

File: spec/controllers/users_controller_spec.rb [hunk 1]
```diff
@@ -4,6 +4,15 @@
 
 RSpec.describe Api::V1::UsersController, type: :controller do
   let(:user) { create(:user) }
+  let(:admin_role) { create(:role) }
```
Comment (KH-Amir-TN) [inline spec/controllers/users_controller_spec.rb:7]: Can we call this something other than admin?
For example `users_manager_role`?

File: spec/controllers/users_controller_spec.rb [hunk 2]
```diff
@@ -68,15 +77,31 @@
     end
   end
 
-  describe 'DELETE users#destroy' do
+  describe 'users#destroy' do
     it 'deletes the user' do
```
Comment (KH-Amir-TN) [inline spec/controllers/users_controller_spec.rb:81]: I think that `deletes the current_user account` is more informational,

File: spec/controllers/users_controller_spec.rb [hunk 3]
```diff
@@ -68,15 +77,31 @@
...
       expect(response).to have_http_status(:ok)
       expect { delete :destroy, params: { id: user.id } }.to change(User, :count).by(-1)
     end
 
     it 'does not delete any user if the user id is invalid' do
```
Comment (KH-Amir-TN) [inline spec/controllers/users_controller_spec.rb:86]: I think that this should be scoped inside of a test group like `context 'current_user has NOT 'ManageUsers' permission'`,
And maybe we can add two tests one for `it 'returns status code :forbidden for a NONE current_user :id '` targeting an existing user and another `it 'returns status code :forbidden for inexistent :id '` targeting an inexistent user,

**Review by KH-Amir-TN — COMMENTED** [review]
Shouldn't we update the `routes.rb`?

File: spec/controllers/users_controller_spec.rb (hunk 1 above)
Comment (scouillard) [reply spec/controllers/users_controller_spec.rb:7]: IMO qualifying an admin as someone with ALL the permissions makes sense.

Then, in negative specs, if you want to test something if the admin does not have a certain permission, you remove it manually.

I've been refactoring those for the past two days and IMO this is the cleaner way to do this.

Comment (scouillard) [reply spec/controllers/users_controller_spec.rb:7]: I have yet to see a case where we need to discriminate between the admin permissions, for example a case where we need to test if an admin has a certain permission but does not have a certain other.

File: spec/controllers/users_controller_spec.rb [hunk 4]
```diff
@@ -68,15 +77,31 @@
...
       expect { delete :destroy, params: { id: 'invalid-id' } }.not_to change(User, :count)
-      expect(response).to have_http_status(:not_found)
+      expect(response).to have_http_status(:forbidden)
+    end
+
+    context 'current_user is an admin' do
```
[the remark this answers is not in the document]
Comment (scouillard) [reply spec/controllers/users_controller_spec.rb:91]: It gets really complicated and messy - unless you find a better solution than I did.

Please look at the whole file.

We create an admin_role at the head of the file.

And we link all the permissions to the admin.

```
  let(:manage_rooms_permission) { create(:permission, name: 'ManageRooms') }
  let!(:manage_rooms_role_permission) do
    create(:role_permission,
           role_id: admin_role.id,
           permission_id: manage_rooms_permission.id,
           value: 'true',
           provider: 'greenlight')
  end

  let(:manage_users_permission) { create(:permission, name: 'ManageUsers') }
  let!(:manage_users_role_permission) do
    create(:role_permission,
           role_id: admin_role.id,
           permission_id: manage_users_permission.id,
           value: 'true',
           provider: 'greenlight')
  end
```

And, in the admin context **before do**, we simply have to set the current_user as an admin with `user.update!(role_id: admin_role.id)`.

Again, after spending two days on this, it's the cleanest way I've found.

**I have yet to see a case where we need to discriminate between the admin permissions, for example a case where we need to test if an admin has a certain permission but does not have a certain other.**

In other words, the user/admin might as well have ALL the permissions - it doesn't change anything.

If needed, you can always manually remove a permission.

**Comment by hadicheaito1** [comment]
The only issue that I see now is that some test files have the case where most of their tests require a permission (such as `roles_controller_spec`). In this case I find it cleaner in a way to just give the user the permission at the beginning and then disable it where needed.

I feel your proposal works best in the cases where majority of the tests in the file don't need the permission being tested.

^ That being said, I'm not sure what the best approach would be atm, but just thought I should share my thinking

**Comment by scouillard** [comment]
> The only issue that I see now is that some test files have the case where most of their tests require a permission (such as `roles_controller_spec`). In this case I find it cleaner in a way to just give the user the permission at the beginning and then disable it where needed.
> 
> I feel your proposal works best in the cases where majority of the tests in the file don't need the permission being tested.
> 
> ^ That being said, I'm not sure what the best approach would be atm, but just thought I should share my thinking

Absolutely - in those cases, you would do the opposite.

In most Specs tho I think this template will work great.

**Comment by scouillard** [comment]
It was obvious after two days of trying to refactor the Specs that something needed to be done for the user "permissions" specs.
I like working with this template, but of course this solution can be improved or even replaced.

File: spec/controllers/users_controller_spec.rb [hunk 5]
```diff
@@ -68,15 +78,38 @@
...
+      expect(response).to have_http_status(:forbidden)
+    end
+
+    it 'returns status code forbidden if the user id is invalid' do
       expect { delete :destroy, params: { id: 'invalid-id' } }.not_to change(User, :count)
-      expect(response).to have_http_status(:not_found)
+      expect(response).to have_http_status(:forbidden)
+    end
+
+    context 'current_user with ManageUsers permission deletes another user' do
```
Comment (KH-Amir-TN) [inline spec/controllers/users_controller_spec.rb:98]: This can be further simplified to something like "current_user with "ManageUsers" permission",

File: spec/controllers/users_controller_spec.rb [hunk 6]
```diff
@@ -68,15 +78,38 @@
...
     end
 
-    it 'does not delete any user if the user id is invalid' do
+    it 'returns status code forbidden if the user tries to delete another user' do
+      new_user = create(:user)
+      expect { delete :destroy, params: { id: new_user.id } }.not_to change(User, :count)
+      expect(response).to have_http_status(:forbidden)
+    end
+
+    it 'returns status code forbidden if the user id is invalid' do
```
Comment (KH-Amir-TN) [inline spec/controllers/users_controller_spec.rb:93]: The word "invalid" may be misleading since we use it to indicate invalid params,
Do you find this more informational "returns status code :forbidden for inexistent :id"?

File: spec/controllers/users_controller_spec.rb [hunk 7]
```diff
@@ -68,15 +78,38 @@
...
       expect { delete :destroy, params: { id: 'invalid-id' } }.not_to change(User, :count)
-      expect(response).to have_http_status(:not_found)
+      expect(response).to have_http_status(:forbidden)
+    end
+
+    context 'current_user with ManageUsers permission deletes another user' do
+      before do
+        manage_users_role_permission
```
Comment (KH-Amir-TN) [inline spec/controllers/users_controller_spec.rb:100]: Is this line required?

Comment (scouillard) [reply spec/controllers/users_controller_spec.rb:100]: Yes, it's needed to instantiate the object.
I would use let! instead but RSpec is complaining that the object is not being directly referenced - maybe we should remove that rule?

Comment (hadicheaito1) [reply spec/controllers/users_controller_spec.rb:100]: Yeah I agree that rule is annoying

File: spec/controllers/users_controller_spec.rb [hunk 8]
```diff
@@ -95,15 +105,47 @@
     end
   end
 
+<<<<<<< HEAD
```
Comment (hadicheaito1) [inline spec/controllers/users_controller_spec.rb:108]: These needs to be removed

File: spec/controllers/users_controller_spec.rb (hunk 6 above)
Comment (scouillard) [reply spec/controllers/users_controller_spec.rb:125]: It is kind of invalid, no?
It's a string and it's expecting an id.
`params: { id: 'invalid-id' }`

**Review by farhatahmad — DISMISSED** [review]

File: spec/controllers/users_controller_spec.rb (hunk 7 above)
Comment (farhatahmad) [reply spec/controllers/users_controller_spec.rb:127]: Yeah I would remove it and use `let!`

File: spec/controllers/users_controller_spec.rb (hunk 6 above)
Comment (KH-Amir-TN) [reply spec/controllers/users_controller_spec.rb:120]: Aren't we testing here the case when we're trying to remove a user that does not exist?
This gives the impression that our API is asserting the data type of the inputs our their values considering some as invalid and returning an errored response.

File: spec/controllers/users_controller_spec.rb [hunk 9]
```diff
@@ -95,15 +105,37 @@
...
+      before do
+        user.update!(role: manage_users_role)
+      end
+
+      it 'deletes a user' do
+        new_user = create(:user)
+        expect { delete :destroy, params: { id: new_user.id } }.to change(User, :count).by(-1)
+      end
+
+      it 'returns status code not found if the user id is invalid' do
```
Comment (KH-Amir-TN) [inline spec/controllers/users_controller_spec.rb:135]: I think this will clearify what I'm trying to say,
This is testing a `:not_found` case, an inexistant id isn't invalid.

Comment (scouillard) [reply spec/controllers/users_controller_spec.rb:135]: Right. Agreed. I don't find this test particularly useful neither.

Comment (scouillard) [reply spec/controllers/users_controller_spec.rb:135]: I changed it to `'returns status code not found if the user does not exists'`

**Review by farhatahmad — APPROVED** [review]
